Project03 Peer Review PR

zoechow24 opened the pull request.

zoechow24 committed:
Project03 Peer Review PR

zoechow24 commented:
Group member is @jacque-caldwell

zoechow24 committed:
Add files via upload

zoechow24 committed:
Add files via upload

Commit:
Fix seq_ops.py

jacque-caldwell committed:
first pseudo code ; added numpy to reticulate loading

added first draft of pseudo code
changed package loading to include numpy in reticulate

zoechow24 committed:
Merge pull request #4 from jacque-caldwell/Project03_PR

first pseudo code ; added numpy to reticulate loading

jacque-caldwell committed:
Update README.md

jacque-caldwell committed:
Update project03.Rmd

zoechow24 committed:
Merge pull request #5 from jacque-caldwell/Project03_PR

Updates to README.md and project03.Rmd

jacque-caldwell committed:
Many changes; still with syntax errors

fixed filenames again in browser (need to take .gz off filename)
Tabs changed in GibbsMotifFinder
Completed documentation on supporting functions fuzzy_diff and create_possible_motif
Changed random function to np.random.random_integer(low=,high=,number_of_things_to_return)
Added "-10" to high so create_possible_motif so we don't read past the end of the array 

Played with tab and in function again, changed format of project to use 5 spaces for tabs to make code clearer.
  Moved a coment into our "pseudocoding" section - was part of the assignment text
  Put error message in select_motif (after changing variable names); commented out line with np. use in it.
  Also put error message in GibbsMotifFinder - with np.random function all in it.
  Also put error message in GibbsMotifFinder - at seqs.pop; as it was failing there....  HOWEVER, N should be 5 not one

zoechow24 committed:
Merge pull request #6 from jacque-caldwell/Project03_PR

Many changes; still with syntax errors

Commit:
Debugged Rmd, added pseudocode to README.md

jacque-caldwell committed:
debugged while loop in GibbsMotifFinder 

debugged while loop (needed to be less than not greater than)

jacque-caldwell reviewed (commented):
reviewed merge

jacque-caldwell reviewed (commented):
reviewed merged changes

zoechow24 committed:
Merge pull request #7 from jacque-caldwell/Project03_PR

debugged while loop in GibbsMotifFinder

jacque-caldwell committed:
changed returned pfm to a npArray
updated file with bug fixes from yesterday (while loop)

jacque-caldwell committed:
Delete project03.Rmd

zoechow24 committed:
Merge pull request #8 from jacque-caldwell/Project03_PR

changed returned pfm to a npArray

zoechow24 committed:
Merge pull request #9 from jacque-caldwell/patch-1

Delete project03.Rmd

jacque-caldwell committed:
created new probabilistic motif selector, choose_motif function.
Integrated into the GibbsMotifFinder function.
Changed big loop to use only two arrays one for motifs and one for the scores.

Not sure that this runs properly - may have infinite loop going ; will have to confirm ways of getting out of the loop

zoechow24 committed:
Merge pull request #10 from jacque-caldwell/Project03_PR

created new probabilistic motif selector, choose_motif function.

Commit:
V3 for GibbsMotifFinder, Update README

jacque-caldwell committed:
* added changes needed to make seqlogo stuff work
* moved imports to top of file (math, seqlogo)

will need to create directory "stuff" in build directory so that seqlogo can put motif.png in it.

zoechow24 committed:
Delete Project03/project03.zip

zoechow24 committed:
Add files via upload

jacque-caldwell committed:
added convert_exp to normalize matrices.
fixed positive and negative DNA strands
fixed a couple of syntax errors

zoechow24 committed:
Merge pull request #12 from jacque-caldwell/Project03_PR

* added changes needed to make seqlogo stuff work

jacque-caldwell committed:
merging by hand

zoechow24 committed:
Merge pull request #13 from jacque-caldwell/Project03_PR

merging by hand

Commit:
changed pfm for IC convergence, edited convert_exp() and removed extra GibbsMotifFinder()

Commit:
Cleaned up Project03.Rmd, added pseudocode to README

Commit:
tested new way to handle complementary strands

jacque-caldwell committed:
output file from our GibbsMotifFinder
adding to project so it can be seen in our Project03.Rmd file.

jacque-caldwell committed:
Fixed a "10 being used intstead of k bug in choose_motif
changed some code comments, deleted testing code and some commented out code.

This code got AGGAGG!!!

jacque-caldwell committed:
Added my reflections in to README.md

zoechow24 committed:
Merge pull request #14 from jacque-caldwell/Project03_PR

completion of project files.

zoechow24 committed:
Update README.md

Commit:
edited variable names

zoechow24 committed:
Update README.md

zoechow24 committed:
Update README.md

zoechow24 committed:
Update README.md

zoechow24 committed:
Update README.md

zoechow24 committed:
Update README.md

zoechow24 committed:
Update README.md

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+      x,y (floats):  values to be compared, are they within epislon of each other if so function returns TRUE, otherwise, returns FALSE
+    epsilon (float) - the amount of difference, or tolerance between X & Y that we will allow
+    
+    Returns (boolean): if abs(x-y) os greater than epislon
```
I think if you intend to return a boolean you should have the function return `True` or `False`. You probably don't want to have unintended consequences of having the return value be an integer (potential source of bugs later, also it's inconsistent with the docs). I do like that you break your logic into functions, but in this specific case, you're essentially wrapping native Python behavior into a function. The expression in the function will already return T/F.

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+    returns a list of nornmalized floats (probabilities)
+  '''
+  
+  exp_scores = [math.exp2(score) for score in poss_scores]
```
if `poss_scores` is cast to `np.ndarray` you can also do `np.exp2` on the entire array at once without having to loop

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+      return(picked_item)
+    else:
+      print(f"Error: choose_motif(): {motif_list}, {current_ic} are not the same length")
+  else:
```
Small stylistic comment - I don't want to be too nitpicky but I think it's common to remove `else` blocks before `return` in Python. Just mentioning this because I noticed a couple of instances of this pattern in the code. You can also do something like `return x if some_expression else y`. All of this makes no difference from a performance standpoint though so it's not a big deal.

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+      j+=1
+      
+    #end of while loop; if we have reached here, there was no convergence return the last motif checked
+    return (np.array(pfm))
```
`build_pfm` is already returning `np.ndarray` so casting it isn't really doing anything

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+      picked_item = random.choices(motif_list, weights, k=1)[0]  
+      return(picked_item)
+    else:
+      print(f"Error: choose_motif(): {motif_list}, {current_ic} are not the same length")
```
If you're intending to catch an error with this, you might want to actually `raise ValueError` or use try/except rather than just printing -- e.g. you can check at the start of the function with something like `if not motif_list or len(motif_list) != len(weights) ... raise ValueError` (or you can do it separately too)

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+      pfm = build_pfm(all_motifs, k)      
+      current_ic = pfm_ic(pfm)
+        
+      if (fuzzy_diff (current_ic,old_ic, 0.005)):
```
Honestly I'm realizing how much I overcomplicated my approach by looking at this - this is very simple and makes more sense than what I was doing. Initially I was trying to calculate the absolute mean or variance difference across iterations, but this is probably easier.

Candy-Train replied:
I agree, this seems like a good, quick way to see "are these close enough"

michael-bambha commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+    """
+          
+          
+    N = random.randint(0, len(sequence)-k)
```
Something I found out is that you can use `np.random.randint()` and return an array of random numbers too! Then you don't need to loop through each sequence.

abimckinley replied:
I was going to say this as well!

abimckinley commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+    if (abs(x-y) < epsilon):
+      return(1) # basically the same!
+    else:
+      return(0) # not the same!
```
Just a teeny thing, but you could also make this return "True" or "False" rather than "1" or "0". 

Using `return abs(x-y) < epsilon` should return True/False.

Tiange commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+        poss_motifs.append(rev_motif)
+        rev_score = score_kmer(rev_motif, bgPWM)
+        poss_motif_scores.append(rev_score)
+        
```
Issue: The code currently appends both the forward and reverse complement k-mers as separate, independent candidates to poss_motifs. A more standard and biologically sound approach is to consider each starting position i as a single candidate and use the best possible score from that position (whether it's on the forward or reverse strand).

Recommendation: For each position i, calculate the scores for both the forward and reverse complement k-mers, but only add the single highest score to your poss_motif_scores list and keep track of the corresponding k-mer string that produced that score.

Tiange commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+      # remove at the end
+      pfm = build_pfm(all_motifs, k)      
+      current_ic = pfm_ic(pfm)
+        
```
pfm = build_pfm(all_motifs, k) and current_ic = pfm_ic(pfm).
These lines are currently placed before the main motif list is updated with the newly selected motif (all_motifs[N] = selected_motif).

Candy-Train commented on Project03/README.md:
```diff
@@ -0,0 +1,181 @@
+To select a new position $m$ (as defined below) use random.choices() or
+numpy.random.choice()
+
+# Pseudocode
```
I really like the pseudocode here. It clearly outlines your plans for each function and what you expect them to do.

Candy-Train commented on Project03/project03.Rmd:
```diff
@@ -0,0 +1,292 @@
+
+
+
+def normalized_scores( poss_scores ):
```
Realizing we needed to normalize the scores for the probability was something my group took a little while to come to. We were having issues trying to feed the score directly into `random.choices` to use for weighted selection.

Candy-Train reviewed (approved):
Looks good to me. Clear functions with singular jobs that align with the plan laid out in the pseudocode.
